src/sampler.jl

```diff
@@ -57,15 +57,61 @@ function sampler_for_process(sampler::TemperedSampler, state::TemperedState, I..
 end
 
 """
-    tempered(sampler, inverse_temperatures; kwargs...)
+    init_internal_sampler(
+        proto_sampler::Union{<:Function,Vector{<:AbstractMCMC.AbstractSampler},AbstractMCMC.AbstractSampler},
+        inverse_temperatures::Union{Integer, Vector{<:Real}};
+        model=nothing,
+        shared_sampler::Bool=false
+    )
+
+Returns an `internal_sampler` depending on the nature of `proto_sampler`, which can take one of the following forms:
+  - a single `sampler::AbstractMCMC`
+  - a list of `n` (== number of inverse temperatures) `samplers::Vector{AbstractMCMC.AbstractSampler}`
+  - a partial `Function` of `model` capable of returning an `AbstractSampler`, i.e. `model -> create_sampler(model, ...)`,
+  note that you must provide the `model` kwarg to the call of `tempered` / `tempered_sample` when using this option
+"""
+function init_internal_sampler(
+    sampler_template::Function,
+    inverse_temperatures::Union{Integer, Vector{<:Real}};
+    model=nothing,
+    kwargs...
+)
+    !isnothing(model) || error("You must provide the `model` kwarg to `tempered` when providing a template function for sampler initialisation.")
+    return [sampler_template(MCMCTempering.make_tempered_model(model, β)) for β in inverse_temperatures]
+end
+
+function init_internal_sampler(
+    sampler::AbstractMCMC.AbstractSampler,
+    inverse_temperatures::Union{Integer, Vector{<:Real}};
+    shared_sampler::Bool=false,
+    kwargs...
+)
+    return shared_sampler ? sampler : [deepcopy(sampler) for _ in eachindex(inverse_temperatures)]
```

Comment:
This shouldn't really be needed. `AbstractSampler` shouldn't be mutated during the sampling process. I say we keep it for now, but maybe add a `# NOTE` something for the future to re-evaluate. For example, I can imagine scenarios where the sampler contains some large pre-computed matrix and this `deepcopy` will then be unfortunate.

```suggestion
    # TODO: Do we really need the `deepcopy` here? `sampler` shouidn't be mutated.
    return shared_sampler ? sampler : [deepcopy(sampler) for _ in eachindex(inverse_temperatures)]
```

Reply:
In your main message, you mention re-adapting the sampler, is this not the precise reason we want to deepcopy it, s.t. each sampler can be adapted separately?